axon-server-connector/src/main/java/org/axonframework/axonserver/connector/command/GrpcBackedCommandMessage.java

```diff
@@ -16,98 +16,85 @@
-        this.request = command;
-        this.serializer = serializer;
+        this.command = command;
+        this.serializedPayload =
+                new LazyDeserializingObject<>(new GrpcSerializedObject(command.getPayload()), serializer);
+        this.metaDataSupplier = new GrpcMetaData(command.getMetaDataMap(), serializer);
+    }
+
+    private GrpcBackedCommandMessage(Command command,
+                                     LazyDeserializingObject<C> serializedPayload,
+                                     Supplier<MetaData> metaDataSupplier) {
+        this.command = command;
+        this.serializedPayload = serializedPayload;
+        this.metaDataSupplier = metaDataSupplier;
     }
 
     @Override
     public String getCommandName() {
-        return request.getName();
+        return command.getName();
     }
 
     @Override
     public String getIdentifier() {
-        return request.getMessageIdentifier();
+        return command.getMessageIdentifier();
     }
 
     @Override
     public MetaData getMetaData() {
-        if (metaData == null) {
-            metaData = deserializeMetaData(request.getMetaDataMap());
-        }
-        return metaData;
+        return metaDataSupplier.get();
     }
 
     @Override
     public C getPayload() {
-        String revision = request.getPayload().getRevision();
-        SerializedObject serializedObject = new SimpleSerializedObject<>(
-                request.getPayload().getData().toByteArray(),
-                byte[].class,
-                request.getPayload().getType(),
-                "".equals(revision) ? null : revision
-        );
-        //noinspection unchecked
-        return (C) serializer.deserialize(serializedObject);
+        return serializedPayload.getObject();
     }
 
     @Override
     public Class<C> getPayloadType() {
-        try {
-            //noinspection unchecked
-            return (Class<C>) Class.forName(request.getPayload().getType());
-        } catch (ClassNotFoundException e) {
-            throw new RuntimeException(e);
-        }
+        return serializedPayload.getType();
     }
 
     @Override
-    public CommandMessage<C> withMetaData(Map<String, ?> metaData) {
-        return this;
+    public GrpcBackedCommandMessage<C> withMetaData(Map<String, ?> metaData) {
+        return new GrpcBackedCommandMessage<>(command, serializedPayload, () -> MetaData.from(metaData));
     }
 
     @Override
-    public CommandMessage<C> andMetaData(Map<String, ?> metaData) {
-        return this;
-    }
-
-    private MetaData deserializeMetaData(Map<String, MetaDataValue> metaDataMap) {
-        if (metaDataMap.isEmpty()) {
-            return MetaData.emptyInstance();
-        }
-        GrpcMetaDataConverter grpcMetaDataConverter = new GrpcMetaDataConverter(serializer);
-        return MetaData.from(grpcMetaDataConverter.convert(metaDataMap));
+    public GrpcBackedCommandMessage<C> andMetaData(Map<String, ?> metaData) {
```

Comment:
Can be moved as a default implementation in the interface Message

Reply:
I catch your drift, but I am a little doubtful to do this :thinking: 

Almost all other implementations of `Message` in our code base also extend from the `MessageDecorator` class.
Due to this, the default solution for the `andMetaData()` calls is:

`new SomeMessageImpl(getDelegate().andMetaData(), ...)`

I feel it might cause some confusion, potentially. But maybe I am to cautious in this case.